### DISCUSSION TITLE
Welcome to WebGrid Discussions!

### DISCUSSION THREAD

**Discussion by TilBlechschmidt** [TilBlechschmidt/WebGrid#11] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Share how you are using WebGrid 🙂
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by madhank93** [comment]
Hey @TilBlechschmidt 
Thanks for developing and maintaining this awesome project. Recently I started to explore scalability in Selenium. As you mentioned in the faq aerokube solution moon is a commercial. And it is not recommend to use Selenoid internally because it happens to stores a list of running sessions in memory.

And then I came across this https://github.com/alcounit/selenosis project and it is stateless. They seems to use sidecar pattern to replaces original sessionId. So I guess even if a pod dies, new pod will spin up and pick up the test from where it left. How does this works in webgrid ?

I am still relatively new to this whole k8s concepts. Please let know if I got it wrong.

**Reply by TilBlechschmidt** [reply]
Hey @madhank93
I'm glad you like the project! It is great to see more people discovering it 🙂 

I have come across Selenosis too but haven't quite managed to gain a complete understanding of the architecture they use.

However, in general, the underlying protocol used by Selenium is stateful. You request a session from some backend (that could be your local browser, a Selenium Hub, WebGrid, you name it) and it in return provides you with an identifier. In later requests this identifier is used to access "your" session until it is terminated.

In that regard, Selenium is and will always be stateful — so will be all intermediary implementations, like WebGrid, that relay requests to sessions. The reason why sessions are not stateless is that you are interacting with browsers by sending a list of requests. For example:

1. Open DuckDuckGo
2. Find the search field HTML element
3. Input some text
4. Press the search button
5. Query for result elements

If your browser/session would die at any point between any of those individual requests, it would be practically impossible to just spin up a new browser, replace the old one and pick up the workflow. Especially, since most operations are not idempotent and thus replaying them is not a possibility. For this reason, to my knowledge, any Selenium Grid equivalent does not handle crashes of browsers (apart from reporting it to the client).

The "stateless" aspect is related to the way the native Selenium Grid works. There, you spin up a fixed number of Browsers and register them at a central hub. This hub then delegates incoming requests to any one of those browsers (whichever one is currently unoccupied)<sup>1</sup>. In most other "_stateless_" implementations, including WebGrid, Selenoid, Moon, and presumably also Selenosis this is not the case. Browsers are dynamically spun up and down as request demands. There is no "state" of registered running browsers. However, there is a state of currently running sessions (browser instances in use by clients)!

I hope this helps in answering your question.

<sub><sup>1</sup>In actuality, every time a client has finished using a browser, it will be torn down and once a new client appears, a new one will be started. This is done to prevent any "leakage" of information or state from one session into another.</sub>

**Reply by TilBlechschmidt** [reply]
What might help is to consider that there are three distinct resources one has to consider:

1. Controllers running on a given machine/pod/container that are responsible for interacting with WebDriver binaries and translating requests
2. Web Browser instances that actually provide the browser itself and render all the UI
3. WebDriver sessions which allow clients to take control of a browser

Selenium Hub statically allocates "Controllers" (they call them "nodes" IIRC) and web browser instances but dynamically create and destroy WebDriver sessions. The controllers register themselves at the Hub and report the available resources on the machine they run on. Since each controller has to be created "manually" and registered at the Hub out of its own accord, dynamic response to incoming requests is not possible.

This has sprouted a number of "alternative" solutions, including WebGrid, which go a step further and dynamically start controller instances as well. It has the benefit of using less resources while not serving any requests (since you do not need constantly running software multiplied by the maximum number of browser sessions you would want to serve at peak usage) and allowing for scalability as you are, theoretically, not limited to any number of maximum controllers. However, it comes with the drawback of needing a bit more time for startup (although, we are talking about five seconds or so).

**Reply by madhank93** [reply]
Thank for the quick response and explaining it very briefly 🙏

**Reply by TilBlechschmidt** [reply]
I hope it helps! It is a bit complicated to explain all the intricate details without going into too much detail and the wording is somewhat ambiguous since "stateless" (among other words) does not necessarily mean the same thing in this context and between projects. If you have any questions or need clarification feel free to ask 🙂

**Reply by madhank93** [reply]
It really helps 🙂. You have made it crystal clear. In my free time I'll try experimenting with webgrid. If I hit any road blocks you will be the first to know.

**Reply by TilBlechschmidt** [reply]
Just as a heads-up: The project is still in a state I'd consider early beta. It definitely works and we've had success running it with multiple hundreds of browsers and under daily load. However, the implementation is still very much in flux and the details need some ironing out. As a matter of fact, just today a major overhaul of the internal code architecture with a focus on maintainability and testability [has been pushed](https://github.com/TilBlechschmidt/WebGrid/tree/feature/architecture-rework) (still a bit buggy here — I'm working on fixing bugs in that codebase as we speak). Once that is done, the documentation should improve further too as things stabilise on the road to v1.0 😉

In other words: Do not rely on it for anything business critical just yet 🙈
